core/src/main/java/org/infinispan/protostream/impl/SerializationContextImpl.java

```diff
@@ -492,6 +521,32 @@ public <T> BaseMarshallerDelegate<T> getMarshallerDelegate(String typeName) {
       }
    }
 
+   @Override
+   public <T> BaseMarshallerDelegate<T> getMarshallerDelegate(int typeId) {
+      // Try first without locking as this class is thread safe
+      Registration registration = marshallersByTypeId.get(typeId);
+      if (registration != null) {
+         return (BaseMarshallerDelegate<T>) registration.marshallerDelegate;
+      }
+
+      // If the id wasn't found we have to acquire the read lock to ensure no concurrent updates
+      long stamp = manifestLock.readLock();
+      try {
+         registration = marshallersByTypeId.get(typeId);
```

Comment:
Why do you need to do the lookup in `marshallersByTypeId` again?

Reply:
It was for concurrent updates, but now that I made the SmallIntMap fully thread safe (to avoid the lock completely), it _probably_ isn't needed. I will mull over it a bit and update.